README.md
# Color_Hue_N_Tint
> This is package to generate color shading palette with hue and tint value from [HSL](https://color.lukas-stratmann.com/color-systems/hsl.html) cylinder.

![Web Testing](https://colorhuentint.web.app/)

![HSL](https://github.com/athiruj/color_hue_n_tint/blob/master/tutorials_img/HSL.png?raw=true)

## Constructor
### HueNTint(int value)
 Color Construct is from the lower 32 bits

```dart
HueNTint color = HueNTint(0xff0000) // Red
```

### HueNTint(int r, int g, int b)
 Construct a color from the lower 8 bits of four integers.
   * `r` is red, from 0 to 255.
   * `g` is green, from 0 to 255.
   * `b` is blue, from 0 to 255.

```dart
HueNTint color = HueNTint(00, 255, 00) // Green
```

## Features
 - #### [.palette()]() - function to create a palette with shading in form of a `List<Color>`.
 - #### [.tints()]() - function to create a light palette in form of a `List<Color>`.
 - #### [.hues()]() - function to create a dark palette in form of a `List<Color>`.
 
### .palette({double? scale})
function to create a palette with shading in form of a `List<Color>`.
 
> ![](https://github.com/athiruj/color_hue_n_tint/blob/master/tutorials_img/Shader_default.png?raw=true) 

```dart
// HueNTint(int value, shades: 10, index: 5, fullScale: false)
HueNTint color = HueNTint(0xff0000);
List<Color> colorList = color.palette(); 
//  output:[
//          Color(0xffffd4d4),
//          Color(0xffffaaaa),
//          Color(0xffff7f7f),
//          Color(0xffff5555),
//          Color(0xffff2a2a),
//          Color(0xffff0000),
//          Color(0xffcc0000),
//          Color(0xff990000),
//          Color(0xff660000),
//          Color(0xff330000)
//         ]

```

### .tints({double? scale})
function to create a light part of the palette.
 
> ![](https://github.com/athiruj/color_hue_n_tint/blob/master/tutorials_img/lightPalette_default.png?raw=true) 

```dart
// HueNTint(int value, shades: 10, index: 5, fullScale: false)
HueNTint color = HueNTint(0xff0000);
List<Color> colorList = color.lightPalette(); 
//  output:[
//          Color(0xffffd4d4),
//          Color(0xffffaaaa),
//          Color(0xffff7f7f),
//          Color(0xffff5555),
//          Color(0xffff2a2a),
//         ]

```

### .hues({double? scale})
function to create a dark part of the palette.

 > ![](https://github.com/athiruj/color_hue_n_tint/blob/master/tutorials_img/darkPalette_default.png?raw=true) 

 ```dart
// HueNTint(int value, shades: 10, index: 5, fullScale: false)
HueNTint color = HueNTint(0xff0000);
List<Color> colorList = color.tints(); 
//  output:[
//          Color(0xffcc0000),
//          Color(0xff990000),
//          Color(0xff660000),
//          Color(0xff330000)
//         ]

```

## Source
 - [The Three Components of Color](https://www.virtualartacademy.com/three-components-of-color/)
 - [HSL](https://color.lukas-stratmann.com/color-systems/hsl.html)
 - [HSL cylinder](https://en.wikipedia.org/wiki/HSL_and_HSV)


CHANGELOG.md
## 1.0.0

- Initial version.
- Shader Class.
  - .palette()
  - .lightPalette()
  - .darkPalette()
  - .lightness()
  - .darkness()
- Extension
  - List\<Color>
    - .toListInt()
    - .toListString  

## 1.0.1

- Change name
  - Shader Class to HueNTint Class
    - .lightPalette() to .tints()
    - .darkPalette() to .hues()
    - .lightness() - delete
    - .darkness() - delete
- Bug fix

## 1.0.2

- Example Update


lib/src/color_hue_n_tint_base.dart
part of color_hue_n_tint;

class HueNTint {
  final int value;

  /// number of color shades
  int _shades = 10;

  /// position of main color in palette
  int _index = 5;

  /// Black White safety
  bool _fullScale = false;

  double _scale = 1.0;

  /// Normaly Color Construct is from the lower 32 bits
  /// but [HueNTint] not support `opcity` or `alpha value` in hex
  /// so [HueNTint] Construct is from the lower 24 bits
  ///
  /// The bits are interpreted as follows:
  ///
  /// * ❌ Bits 24-31 are the alpha/opcity value.
  /// * ✅ `Bits 16-23 are the red value.`
  /// * ✅ `Bits 8-15 are the green value.`
  /// * ✅ `Bits 0-7 are the blue value.`
  ///
  /// In other words, if 'RR' the red value in hex, 'GG' the green value in hex,
  /// and 'BB' the blue value in hex, a color can be
  /// expressed as `const ColorShade(0xRRGGBB)`.
  ///
  /// **parameter:**
  /// * **int shades**
  ///   number of Color in palette.
  ///
  /// * **int index**
  ///   position of main Color in palette.
  ///
  /// * **bool fullScale**
  ///   full minimum & maximum value of scale 0 -> 255
  ///   * `fullScale: false`
  ///   Black White Safety
  /// ![Shader_default](https://github.com/athiruj/color_hue_n_tint/blob/master/tutorials_img/Shader_default.png?raw=true)
  ///   * `fullScale: true`
  ///   Full scale with Black White
  /// ![Shader_full](https://github.com/athiruj/color_hue_n_tint/blob/master/tutorials_img/Shader_full.png?raw=true)
  ///
  HueNTint(int value, {int? shades, int? index, bool fullScale = false})
      : value = value & 0xffffffff {
    assert(_shades >= _index, 'shades($_shades) >= index($_index)');
    assert(_shades >= 0, 'Shades($_shades) can\'s >= 0');
    _shades = shades ?? _shades;
    _index = index ?? _index;
    _fullScale = fullScale;
  }

  /// Construct a color from the lower 8 bits of four integers.
  ///
  /// * `r` is [red], from 0 to 255.
  /// * `g` is [green], from 0 to 255.
  /// * `b` is [blue], from 0 to 255.
  ///
  /// Out of range values are brought into range using modulo 255.
  ///
  /// See also [fromRGB], which takes the alpha value as a floating point
  /// value.
  ///
  /// For example, `const HueNTint(255, 56, 211)`.
  /// `255` is the Red value in hex,
  ///  `56` is the Green value in hex,
  /// and `211` is the Blue value in hex.
  HueNTint.fromRGB(int r, int g, int b,
      {int? shades, int? index, bool fullScale = false})
      : value = (0xff000000 | (r << 16) | (g << 8) | (b << 0)) & 0xffffffff {
    assert(_shades >= _index, 'shades($_shades) >= index($_index)');
    assert(_shades >= 0, 'Shades($_shades) can\'s >= 0');
    assert(r > 0 && r > 255, 'red value from 0 - 255');
    assert(g > 0 && g > 255, 'green value from 0 - 255');
    assert(b > 0 && b > 255, 'blue value from 0 - 255');
    _shades = shades ?? _shades;
    _index = index ?? _index;
    _fullScale = fullScale;
  }

  /// `[r]` The red channel of this color in an 8 bit value.
  int get r => (0x00ff0000 & value) >> 16;

  /// `[g]` The red channel of this color in an 8 bit value.
  int get g => (0x0000ff00 & value) >> 8;

  /// `[b]` The red channel of this color in an 8 bit value.
  int get b => (0x000000ff & value) >> 0;

  int get shades => _shades;

  bool get fullScale => _fullScale;

  int get index => _index;

  int get lightShades => _index;

  int get darkShades => _shades - (_index + 1);

  List<double> lightScale() {
    int s = fullScale ? 0 : 1;
    return [
      double.parse(((255 - r) / (lightShades + s) * _scale).toStringAsFixed(2)),
      double.parse(((255 - g) / (lightShades + s) * _scale).toStringAsFixed(2)),
      double.parse(((255 - b) / (lightShades + s) * _scale).toStringAsFixed(2))
    ];
  }

  List<double> darkScale() {
    int s = fullScale ? 0 : 1;
    return [
      double.parse((r / (darkShades + s) * _scale).toStringAsFixed(2)),
      double.parse((g / (darkShades + s) * _scale).toStringAsFixed(2)),
      double.parse((b / (darkShades + s) * _scale).toStringAsFixed(2))
    ];
  }

  /// [palette] is function to create a palette with shading in the form of a `List<Color>`.
  ///
  /// **parameter:**
  ///
  /// * double `scale` is `0.0(0%)` to `1.0(100%)` of value per Color , Default is `1.0(100%)`
  ///
  /// **Example:**
  /// * `Color(0xff0000, fullScale: false).palate()`
  ///
  ///   **output:**
  /// ```
  ///   [
  ///       Color(0xffffd4d4),
  ///       Color(0xffffaaaa),
  ///       Color(0xffff7f7f),
  ///       Color(0xffff5555),
  ///       Color(0xffff2a2a),
  ///       Color(0xffff0000),
  ///       Color(0xffcc0000),
  ///       Color(0xff990000),
  ///       Color(0xff660000),
  ///       Color(0xff330000)
  ///    ]
  ///```
  /// ![Shader_defualt](https://github.com/athiruj/color_hue_n_tint/blob/master/tutorials_img/Shader_default.png?raw=true)
  ///
  List<Color> palette({double? scale}) {
    _scale = scale ?? _scale;
    List<Color> l = tints(scale: _scale);
    List<Color> d = hues(scale: _scale);
    return l.reversed.toList() + [Color(0xff000000 | value & 0xffffffff)] + d;
  }

  /// [tints] is function to create a light palette in the form of a `List<Color>`.
  ///
  /// **parameter:**
  ///
  /// * double `scale` is `0.0(0%)` to `1.0(100%)` of value per Color , Default is `1.0(100%)`
  ///
  /// **Example:**
  /// * `HueNTint(0xff0000, fullScale: false).tints()`
  ///
  ///   **output:**
  /// ```
  ///   [
  ///       Color(0xffffd4d4),
  ///       Color(0xffffaaaa),
  ///       Color(0xffff7f7f),
  ///       Color(0xffff5555),
  ///       Color(0xffff2a2a),
  ///    ]
  ///```
  /// ![lightPalette](https://github.com/athiruj/color_hue_n_tint/blob/master/tutorials_img/lightPalette_default.png?raw=true)
  ///
  List<Color> tints({double? scale}) {
    _scale = scale ?? _scale;
    assert(_scale >= 0.0 && _scale <= 1.0, '0.0 >= scale <= 1.0');
    return List.generate(
        lightShades,
        (n) => Color(0xff000000 |
            ((r + (lightScale()[0] * _scale * (n + 1))).toInt() << 16) |
            ((g + (lightScale()[1] * _scale * (n + 1))).toInt() << 08) |
            ((b + (lightScale()[2] * _scale * (n + 1))).toInt() << 00) &
                0xffffffff));
  }

  /// [hues] is function to create a dark palette in the form of a `List<Color>`.
  ///
  /// **parameter:**
  ///
  /// * double `scale` is `0.0(0%)` to `1.0(100%)` of value per Color , Default is `1.0(100%)`
  ///
  /// **Example:**
  /// * `HueNTint(0xff0000, fullScale: false).hues()`
  ///
  ///   **output:**
  /// ```
  ///   [
  ///       Color(0xffcc0000),
  ///       Color(0xff990000),
  ///       Color(0xff660000),
  ///       Color(0xff330000)
  ///    ]
  ///```
  /// ![darkPalette](https://github.com/athiruj/color_hue_n_tint/blob/master/tutorials_img/darkPalette_default.png?raw=true)
  ///
  List<Color> hues({double? scale}) {
    _scale = scale ?? _scale;
    assert(_scale >= 0.0 && _scale <= 1.0, '0.0 >= scale <= 1.0');
    return List.generate(
        darkShades,
        (n) => Color(0xff000000 |
            ((r - (darkScale()[0] * _scale * (n + 1))).toInt() << 16) |
            ((g - (darkScale()[1] * _scale * (n + 1))).toInt() << 08) |
            ((b - (darkScale()[2] * _scale * (n + 1))).toInt() << 00) &
                0xffffffff));
  }

  /// Shading % of value to White
  // Color lightness({required double? scale}) {
  //   _scale = scale ?? _scale;
  //   return Color(0xff000000 |
  //       ((r + ((255 - r) * _scale)).toInt() << 16) |
  //       ((g + ((255 - g) * _scale)).toInt() << 08) |
  //       ((b + ((255 - b) * _scale)).toInt() << 00) & 0xffffffff);
  // }

  /// Shading % of value to Black
  // Color darkness({required double? scale}) {
  //   _scale = scale ?? _scale;
  //   return Color(0xff000000 |
  //       ((r - (r * _scale)).toInt() << 16) |
  //       ((g - (g * _scale)).toInt() << 08) |
  //       ((b - (b * _scale)).toInt() << 00) & 0xffffffff);
  // }
}
